File: backend/app/schemas/task.py
from datetime import date, datetime
from typing import Optional

from pydantic import BaseModel

from app.schemas.checklist_item import ChecklistItemResponse
from app.schemas.enums import Priority, TaskStatus, TaskType


class TaskCreateRequest(BaseModel):
    title: str
    task_type: TaskType
    category: Optional[str] = None
    priority: Priority
    due_date: Optional[date] = None
    parent_id: Optional[int] = None
    done_criteria: str
    decision_criteria: Optional[str] = None
    reversible: Optional[bool] = None
    exploration_limit: Optional[int] = None


class TaskUpdateRequest(BaseModel):
    title: Optional[str] = None
    task_type: Optional[TaskType] = None
    category: Optional[str] = None
    priority: Optional[Priority] = None
    status: Optional[TaskStatus] = None
    due_date: Optional[date] = None
    done_criteria: Optional[str] = None
    decision_criteria: Optional[str] = None
    reversible: Optional[bool] = None
    exploration_limit: Optional[int] = None


class TaskResponse(BaseModel):
    id: int
    title: str
    task_type: TaskType
    category: Optional[str] = None
    priority: Priority
    status: TaskStatus
    due_date: Optional[date] = None
    parent_id: Optional[int] = None
    done_criteria: str
    decision_criteria: Optional[str] = None
    reversible: Optional[bool] = None
    exploration_limit: Optional[int] = None
    origin_checklist_item_id: Optional[int] = None
    last_updated_at: datetime
    created_at: datetime

    model_config = {"from_attributes": True}


class OriginInfo(BaseModel):
    parent_task_id: int
    parent_task_title: str
    checklist_item_text: str


class TaskDetailResponse(TaskResponse):
    children: list[TaskResponse] = []
    checklist: list[ChecklistItemResponse] = []
    origin: Optional[OriginInfo] = None


class StaleTaskResponse(TaskResponse):
    stale_days: int
    threshold_days: int


class CarryoverCandidateResponse(TaskResponse):
    overdue_days: int


class ConvergenceChecklist(BaseModel):
    options_within_limit: bool
    structure_simplified: bool
    reversible_confirmed: bool


class ConvergenceResponse(BaseModel):
    task_id: int
    exploration_limit: Optional[int] = None
    exploration_used: int
    exploration_remaining: Optional[int] = None
    reversible: Optional[bool] = None
    decision_criteria: Optional[str] = None
    is_convergeable: bool
    convergence_checklist: ConvergenceChecklist

File: frontend/app/types/task.ts
export type TaskType = 'research' | 'decision' | 'execution'
export type TaskStatus = 'todo' | 'doing' | 'done' | 'carryover_candidate' | 'needs_redefine' | 'snoozed'
export type Priority = 'must' | 'should'

export interface Task {
  id: number
  title: string
  task_type: TaskType
  category: string | null
  priority: Priority
  status: TaskStatus
  due_date: string | null
  parent_id: number | null
  done_criteria: string
  decision_criteria: string | null
  reversible: boolean | null
  exploration_limit: number | null
  origin_checklist_item_id: number | null
  last_updated_at: string
  created_at: string
}

export interface TaskDetail extends Task {
  children: Task[]
  checklist: ChecklistItem[]
  origin: OriginInfo | null
}

export interface OriginInfo {
  parent_task_id: number
  parent_task_title: string
  checklist_item_text: string
}

export interface ChecklistItem {
  id: number
  task_id: number
  text: string
  is_done: boolean
  order_no: number
  extracted_task_id: number | null
}

export interface TaskCreateRequest {
  title: string
  task_type: TaskType
  category?: string | null
  priority: Priority
  due_date?: string | null
  parent_id?: number | null
  done_criteria: string
  decision_criteria?: string | null
  reversible?: boolean | null
  exploration_limit?: number | null
}

export interface TaskUpdateRequest {
  title?: string
  task_type?: TaskType
  category?: string | null
  priority?: Priority
  status?: Exclude<TaskStatus, 'done'>
  due_date?: string | null
  done_criteria?: string
  decision_criteria?: string | null
  reversible?: boolean | null
  exploration_limit?: number | null
}

export interface TaskQueryParams {
  status?: TaskStatus
  task_type?: TaskType
  priority?: Priority
  parent_id?: number | null
}

export interface ConvergenceChecklist {
  options_within_limit: boolean
  structure_simplified: boolean
  reversible_confirmed: boolean
}

export interface ConvergenceInfo {
  task_id: number
  exploration_limit: number | null
  exploration_used: number
  exploration_remaining: number | null
  reversible: boolean | null
  decision_criteria: string | null
  is_convergeable: boolean
  convergence_checklist: ConvergenceChecklist
}
